sotodlib/site_pipeline/make_position_match.py

```diff
@@ -0,0 +1,1014 @@
+
+        cpd_args: Dictionairy containing kwargs to be passed into AffineRegistration.
+                  See the pycpd docs for what these can be.
+    Returns:
+
+        mapping: Mapping between elements in template and focal_plane.
+                 focal_plane[i] = template[mapping[i]]
+
+        outliers: Indices of points that are outliers.
+                  Note that this is in the basis of mapping and focal_plane, not template.
+
+        P: The likelihood array without priors applied.
+
+        TY: The transformed points.
+    """
+    if not bias_lines:
+        focal_plane = focal_plane[:, 1:]
+        template = template[:, 1:]
+    if reverse:
+        cpd_args.update({"X": focal_plane, "Y": template})
+    else:
+        cpd_args.update({"Y": focal_plane, "X": template})
+    reg = AffineRegistration(**cpd_args)
+
+    if vis:
+        frames = []
+
+        def store_frames(frames, iteration, error, X, Y):
+            frames += [[iteration, error, X, Y]]
+
+        fig = plt.figure()
+        fig.add_axes([0, 0, 1, 1])
+        callback = partial(store_frames, frames=frames)
+        reg.register(callback)
+        anim = ani.FuncAnimation(
+            fig=fig,
+            func=partial(
+                visualize, frames=frames, ax=fig.axes[0], bias_lines=bias_lines
+            ),
+            frames=len(frames),
+            interval=200,
+        )
+        plt.show()
+    else:
+        reg.register()
+
+    P = reg.P.T
+    if reverse:
+        P = reg.P
+    TY = reg.TY
+    if not bias_lines:
+        TY = np.column_stack((np.zeros(len(TY)), TY))
+
+    if priors is None:
+        priors = 1
+
+    # Solve the assignment problem
+    row_ind, col_ind = linear_sum_assignment(P * priors, True)
+    if len(row_ind) < len(focal_plane):
+        mapping = np.argmax(P * priors, axis=0)
+        mapping[col_ind] = row_ind
+    else:
+        mapping = row_ind[np.argsort(col_ind)]
+
+    outliers = np.array([])
+    if out_thresh > 0:
+        outliers = np.where(P[mapping, range(P.shape[1])] < out_thresh)[0]
+
+    return mapping, outliers, P, TY
+
+
+def _gen_template(ufm):
+    logger.info("Generating template for " + ufm)
+    wafer = MapMaker(north_is_highband=False, array_name=ufm)
+    det_x = []
+    det_y = []
+    polang = []
+    det_ids = []
+    template_bg = []
+    is_north = []
+    for det in wafer.grab_metadata():
+        if not det.is_optical:
+            continue
+        det_x.append(det.det_x)
+        det_y.append(det.det_y)
+        polang.append(det.angle_actual_deg)
+        det_ids.append(det.detector_id)
+        template_bg.append(det.bias_line)
+        is_north.append(det.is_north)
+    det_ids = np.array(det_ids)
+    template_bg = np.array(template_bg)
+    template_msk = np.isin(template_bg, valid_bg)
+    template_n = np.array(is_north) & template_msk
+    template_s = ~np.array(is_north) & template_msk
+    template_bg[template_bg % 2 == 1] *= -1
```

Comment:
You need to write a comment in the code about why you're doing this. :)

One self-documenting way is to define a function, _scramble_bgs(bgs), and explain in there what the point of this is.  Then everywhere you need to do this, it's like template_bg = _scramble_bgs(template_bg) and an editor can much more easily find/check all the places this is being done.